third_party/prost/private/BUILD.bazel

```diff
@@ -0,0 +1,79 @@
+load("@rules_proto//proto:defs.bzl", "proto_lang_toolchain")
+load("@rules_rust//rust:defs.bzl", "rust_binary", "rust_library_group")
+load(":prost.bzl", "PROST_EXTENSION", "TONIC_EXTENSION", "RUST_EDITION")
+
+rust_library_group(
+    name = "prost_runtime",
+    deps = [
+        "//third_party/prost/private/3rdparty/crates:prost",
+        "//third_party/prost/private/3rdparty/crates:prost-types",
+    ],
+)
+
+proto_lang_toolchain(
+    name = "prost_proto_toolchain",
+    command_line = "--prost_out=$(OUT)",
+    mnemonic = "ProstGenProto",
+    plugin = "//third_party/prost/private/3rdparty/crates:protoc-gen-prost__protoc-gen-prost",
+    plugin_format_flag = "--plugin=protoc-gen-prost=%s",
+    proto_compiler = ":prost_protoc",
+    runtime = ":prost_runtime",
+    tags = ["manual"],
+    visibility = ["//visibility:public"],
+)
+
+rust_library_group(
+    name = "tonic_runtime",
+    deps = [
+        "//third_party/prost/private/3rdparty/crates:prost",
+        "//third_party/prost/private/3rdparty/crates:prost-types",
+        "//third_party/prost/private/3rdparty/crates:tonic",
+    ],
+)
+
+proto_lang_toolchain(
+    name = "tonic_proto_toolchain",
+    # Note that `--tonic_out` is handeld by the proto compiler
+    command_line = "--prost_out=$(OUT)",
+    mnemonic = "TonicGenProto",
+    plugin = "//third_party/prost/private/3rdparty/crates:protoc-gen-prost__protoc-gen-prost",
+    plugin_format_flag = "--plugin=protoc-gen-prost=%s",
+    proto_compiler = ":tonic_protoc",
+    runtime = ":tonic_runtime",
+    tags = ["manual"],
+    visibility = ["//visibility:public"],
+)
+
+[
+    rust_binary(
+        name = "{}_protoc".format(kind),
+        srcs = ["protoc_wrapper.rs"],
+        edition = RUST_EDITION,
+        crate_features = [kind],
+        data = [
+            "@com_google_protobuf//:protoc",
+            "@rules_rust//rust/toolchain:current_rustfmt_files",
+        ] + ([
+            "//third_party/prost/private/3rdparty/crates:protoc-gen-tonic__protoc-gen-tonic",
+        ] if kind == "tonic" else []),
+        rustc_env = dict({
+            "PROST_EXTENSION": PROST_EXTENSION,
+            "PROTOC": "$(rootpath @com_google_protobuf//:protoc)",
```

Comment:
The `protoc` and `protoc-gen-tonic` plugin should probably come from the rule. This process wrapper could stand to be dumber